parsel.ts

```diff
@@ -0,0 +1,475 @@
+export const enum TokenType {
+    Class = 'class',
+    Attribute = 'attribute',
+    Id = 'id',
+    Type = 'type',
+    Universal = 'universal',
+    PseudoElement = 'pseudo-element',
+    PseudoClass = 'pseudo-class',
+    Comma = 'comma',
+    Combinator = 'combinator'
+}
+
+export const TOKENS: Record<string, RegExp> = {
+    [TokenType.Attribute]:
```

Comment:
This seems a little *too* granular and repetitive, are you sure it's needed? What purpose does it serve? What's the advantage over having a regular object literal that's declared as `Record<string, RegExp>`?

Reply:
`TOKENS` without the `Record<string, RegExp>` becomes an object of `Record<TokenType, RegExp>`. You've stated in your docs that others can add their own token types though. This is why I chose `Record<string, RegExp` which allows this.

Reply:
No I mean why not define it as just `Record<string, RegExp>`? What purpose does it serve to restrict the type of the keys like this?

Reply:
Perhaps you're talking about defining the object as `{ attribute: ... }` instead of `{ [TokenType.Attribute]: ... }`? They are both the same, and they end up more restrictive than `Record<string, RegExp>` as they only allow keys defined at definition type. If this is not what you mean, perhaps you can illustrate with code.

Reply:
Alternatively, we could define the key as `any` if restricting the key type to a string is the problem, but it's always better to restrict the type in a min-max sense.

Reply:
Sorry, I have not been explaining what I meant well. I'll try again:

The only typing restriction we need for this object literal is:
a) the keys are strings
b) the values are regexes

We don't need to further restrict the actual values of the keys, since the object is extensible anyway.

It looks like to me that something like this would work to do that (without the `TokenType` enum at all):

```js
export const TOKENS: Record<string, RegExp> = {
	attribute:
		/\[\s*(?:(?<namespace>\*|[-\w]*)\|)?(?<name>[-\w\u{0080}-\u{FFFF}]+)\s*(?:(?<operator>\W?=)\s*(?<value>.+?)\s*(\s(?<caseSensitive>[iIsS]))?\s*)?\]/gu,
	id: /#(?<name>(?:\\.|[-\w\u{0080}-\u{FFFF}])+)/gu,
	class: /\.(?<name>(?:\\.|[-\w\u{0080}-\u{FFFF}])+)/gu,
	comma: /\s*,\s*/g, // must be before combinator
	combinator: /\s*[\s>+~]\s*/g, // this must be after attribute
	"pseudo-element":
		/::(?<name>[-\w\u{0080}-\u{FFFF}]+)(?:\((?<argument>¶+)\))?/gu, // this must be before pseudo-class
	"pseudo-class":
		/:(?<name>[-\w\u{0080}-\u{FFFF}]+)(?:\((?<argument>¶+)\))?/gu,
	universal: /(?:(?<namespace>\*|[-\w]*)\|)?\*/gu,
	type:
		/(?:(?<namespace>\*|[-\w]*)\|)?(?<name>[-\w\u{0080}-\u{FFFF}]+)|\*/gu // this must be last
};
```

I did try it out and it seems to work, but it's entirely possible I'm missing something.

Reply:
In general, the reason to use `enum`s is because they allow you to give a meaningful, programmatic name of a value, independent of the value itself.

Reply:
I think we are on the same page now. Using an enum is just a choice. We can use strings directly if you want, but declaring the enum hints to users of the library that the values of the enum can be used for modifying the `TOKENS` object. Without it, people will have to look into the library code to determine what keys exists since the object is just `Record<string, RegExp>`

Reply:
sgtm